packages/angular/ssr/src/app-engine.ts

```diff
@@ -0,0 +1,156 @@
+
+/**
+ * Angular server application engine.
+ * Manages localized Angular server applications, handles rendering requests,
+ * and optionally transforms index HTML before rendering.
+ */
+export class AngularAppEngine {
+  /**
+   * Hooks for extending or modifying the behavior of the server application.
+   * @internal This property is accessed by the Angular CLI when running the dev-server.
+   */
+  static hooks = new Hooks();
+
+  /**
+   * Hooks for extending or modifying the behavior of the server application.
+   * This instance can be used to attach custom functionality to various events in the server application lifecycle.
+   */
+  get hooks(): Hooks {
+    return AngularAppEngine.hooks;
+  }
+
+  /**
+   * Indicates whether the application is in development mode.
+   * @internal This property is set by the Angular CLI when running the dev-server.
+   */
+  static isDevMode = false;
+
+  /**
+   * The manifest for the server application.
+   * @internal This property is set by the Angular CLI during the build.
+   */
+  static manifest: AngularAppEngineManifest;
+
+  /**
+   * Returns the manifest associated with this server application.
+   * @returns The manifest object.
+   */
+  private get manifest() {
+    return AngularAppEngine.manifest;
+  }
+
+  /**
+   * Map of locale strings to corresponding `AngularServerApp` instances.
+   * Each instance represents an Angular server application.
+   */
+  private readonly appsCache = new Map<string, AngularServerApp>();
+
+  /**
+   * Map of entry points from the `AngularAppEngine` manifest.
+   * Maps locale strings to their respective entry point paths.
+   */
+  private readonly entryPoints = new Map(Object.entries(this.manifest.entryPoints));
+
+  /**
+   * Renders an HTTP request using the appropriate Angular server application and returns a response.
+   *
+   * This method determines the entry point for the Angular server application based on the request URL,
+   * and caches the server application instances for reuse. If the application is in development mode,
+   * the cache is bypassed and a new instance is created for each request.
+   *
+   * If the request URL appears to be for a file (excluding `/index.html`), the method returns `null`.
+   * A request to `https://www.example.com/page/index.html` will render the Angular route
+   * corresponding to `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered.
+   * @param requestContext - Optional additional context for the request, such as metadata.
+   * @returns A promise that resolves to a Response object or null if the request is file-like.
+   */
+  async render(request: Request, requestContext?: unknown): Promise<Response | null> {
+    // Skip if the request looks like a file but not `/index.html`.
+    const url = new URL(request.url);
+    const { pathname } = url;
+    if (isFileLike(pathname) && !pathname.endsWith('/index.html')) {
+      return null;
+    }
```

Comment:
I'm wondering if we should consider returning a rejected promise in case an incorrect/unsupported URL is used or if a request is made for a route that doesn't exist in the manifest (we currently throw an error from the `getEntryPointFromUrl` function).

Reply:
I’ve updated the `getEntryPointFromUrl` method to return `null` for unhandled URLs. This indicates that Angular does not process the URL, allowing the request lifecycle to proceed to the next middleware. 

It's important to distinguish between an unhandled URL and an error, as they represent different scenarios and should follow separate code paths.